File: gui_view/src/main/scala/com/goexp/galgame/gui/view/game/CommonTabController.java
package com.goexp.galgame.gui.view.game;

import com.goexp.galgame.gui.model.Game;
import com.goexp.galgame.gui.task.TaskService;
import com.goexp.galgame.gui.util.FXMLLoaderProxy;
import com.goexp.galgame.gui.view.game.listview.DataViewController;
import javafx.collections.ObservableList;
import javafx.concurrent.Service;
import javafx.concurrent.Task;
import javafx.scene.layout.Region;

import java.net.URL;
import java.util.function.Predicate;
import java.util.function.Supplier;

public class CommonTabController {

    public Region node;
    public DataViewController controller;

    private Service<ObservableList<Game>> gameSearchService;
    private URL resource = DataViewController.class.getResource("dataview.fxml");


    public CommonTabController(Supplier<Task<ObservableList<Game>>> task) {
        this.gameSearchService = new TaskService<>(task);

        init();
    }

    public CommonTabController(Service<ObservableList<Game>> service) {
        this.gameSearchService = service;

        init();
    }

    private void init() {

        final var loader = new FXMLLoaderProxy<Region, DataViewController>(resource);
        node = loader.node;
        controller = loader.controller;

        controller.reloadProperty.addListener((observable, oldValue, newValue) -> {
            if (newValue != null && newValue)
                gameSearchService.restart();
        });

        gameSearchService.valueProperty().addListener((observable, oldValue, newValue) -> {
            if (newValue != null) {

                if (initPredicate != null)
                    controller.load(newValue, initPredicate);
                else
                    controller.load(newValue);
            }
        });

        controller.progessloading.visibleProperty().bind(gameSearchService.runningProperty());

    }


    private Predicate<Game> initPredicate;

    public void load(Predicate<Game> initPredicate) {

        this.initPredicate = initPredicate;
        load();
    }

    public void load() {
        gameSearchService.restart();
    }

}

File: gui_view/src/main/scala/com/goexp/galgame/gui/view/game/CommonTabController.scala
package com.goexp.galgame.gui.view.game

import java.util.function.Predicate

import com.goexp.galgame.gui.model.Game
import com.goexp.galgame.gui.task.TaskService
import com.goexp.galgame.gui.util.FXMLLoaderProxy
import com.goexp.galgame.gui.view.game.listview.DataViewController
import javafx.collections.ObservableList
import javafx.concurrent.Task
import javafx.scene.layout.Region

class CommonTabController(private[this] val taskCreator: () => Task[ObservableList[Game]]) {
  val loader = new FXMLLoaderProxy[Region, DataViewController](classOf[DataViewController].getResource("dataview.fxml"))
  val node = loader.node
  val controller = loader.controller
  private val gameSearchService = TaskService(taskCreator)

  init()


  private def init() = {

    controller.reloadProperty.addListener((_, _, newValue) => {
      if (newValue) gameSearchService.restart()

    })
    gameSearchService.valueProperty.addListener((_, _, newValue) => {
      if (newValue != null)
        if (initPredicate != null) controller.load(newValue, initPredicate)
        else controller.load(newValue)
    })

    controller.progessloading.visibleProperty.bind(gameSearchService.runningProperty)
  }

  private var initPredicate: Predicate[Game] = _

  def load(initPredicate: Predicate[Game]): Unit = {
    this.initPredicate = initPredicate
    load()
  }

  def load(): Unit = gameSearchService.restart()
}